src/order/filter/bases.lean

```diff
@@ -670,6 +677,18 @@ end
 
 end two_types
 
+open equiv
+
+lemma prod_assoc (f : filter α) (g : filter β) (h : filter γ) :
+  map (prod_assoc α β γ) ((f ×ᶠ g) ×ᶠ h) = f ×ᶠ (g ×ᶠ h) :=
+begin
+  rw ← comap_equiv_symm,
+  apply ((((basis_sets f).prod $ basis_sets g).prod $ basis_sets h).comap _).eq_of_same_basis,
+  convert ((basis_sets f).prod (((basis_sets g)).prod
+           (basis_sets h))).comp_equiv (prod_assoc _ _ _) ;
+  { ext, simp [and_assoc] }
+end
```

Comment:
With the `prod_assoc_image` lemma from #12022 this golfs to
```suggestion
begin
  apply ((((basis_sets f).prod $ basis_sets g).prod $ basis_sets h).map _).eq_of_same_basis,
  have := ((basis_sets f).prod $ (basis_sets g).prod $ basis_sets h).comp_equiv (prod_assoc _ _ _),
  simpa only [prod_assoc_image, function.comp, and_assoc] using this,
end
```

Reply:
Indeed those lemmas are natural to extract. Note that `simpa [...] using this` can be compressed to `simpa [...]`.